packages/grafana-ui/src/components/DropZone/DropZone.tsx

```diff
@@ -0,0 +1,216 @@
+import { css, cx } from '@emotion/css';
+import { GrafanaTheme2 } from '@grafana/data';
+import { uniqueId } from 'lodash';
+import React, { ReactNode, useCallback, useState } from 'react';
+import { DropEvent, DropzoneOptions, FileRejection, useDropzone } from 'react-dropzone';
+import { useTheme2 } from '../../themes';
+import { Icon } from '../Icon/Icon';
+import { FileListItem } from './FileListItem';
+
+export interface DropzoneProps {
+  /**
+   * Use the children property to have custom dropzone view.
+   */
+  children?: ReactNode;
+  /**
+   * Use this property to override the default behaviour for the react-dropzone options.
+   * @default {
+   *  maxSize: Infinity,
+   *  minSize: 0,
+   *  multiple: true,
+   *  maxFiles: 0,
+   * }
+   */
+  options?: DropzoneOptions;
+  /**
+   * Use this to change the FileReader's read.
+   */
+  readAs?: 'readAsArrayBuffer' | 'readAsText' | 'readAsBinaryString' | 'readAsDataURL';
+  /**
+   * Use the onLoad function to get the result from FileReader.
+   */
+  onLoad?: (result: string | ArrayBuffer | null) => void;
+}
+
+export interface DropzoneFile {
+  file: File;
+  id: string;
+  error: DOMException | null;
+  progress?: number;
+  abortUpload?: () => void;
+  retryUpload?: () => void;
+}
+
+export function Dropzone({ options, children, readAs, onLoad }: DropzoneProps) {
+  const [files, setFiles] = useState<DropzoneFile[]>([]);
+
+  const setFileProperty = useCallback(
+    (customFile: DropzoneFile, action: (customFileToModify: DropzoneFile) => void) => {
+      setFiles((oldFiles) => {
+        const newFiles = oldFiles.map((f) => {
+          if (f.id === customFile.id) {
+            action(f);
+            return f;
+          }
+          return f;
+        });
+        return newFiles;
+      });
```

Comment:
```suggestion
        return oldFiles.map((f) => {
          if (f.id === customFile.id) {
            action(f);
            return f;
          }
          return f;
        });
      });
```
Doesn't seem like we need to create an extra variable here.

Reply:
true.